config/src/iroha.rs

```diff
@@ -60,26 +62,83 @@ view! {
-        let sumeragi_configuration = sumeragi::Configuration::default();
-        let (public_key, private_key) = sumeragi_configuration.key_pair.clone().into();
-
         Self {
-            public_key,
-            private_key,
-            disable_panic_terminal_colors: bool::default(),
-            kura: kura::Configuration::default(),
-            sumeragi: sumeragi_configuration,
-            torii: torii::Configuration::default(),
-            block_sync: block_sync::Configuration::default(),
-            queue: queue::Configuration::default(),
-            logger: logger::Configuration::default(),
-            genesis: genesis::Configuration::default(),
-            wsv: wsv::Configuration::default(),
-            network: network::Configuration::default(),
-            telemetry: telemetry::Configuration::default(),
+            public_key: None,
+            private_key: None,
+            disable_panic_terminal_colors: Some(bool::default()),
+            kura: Some(kura::ConfigurationProxy::default()),
+            sumeragi: Some(sumeragi::ConfigurationProxy::default()),
+            torii: Some(torii::ConfigurationProxy::default()),
+            block_sync: Some(block_sync::ConfigurationProxy::default()),
+            queue: Some(queue::ConfigurationProxy::default()),
+            logger: Some(logger::ConfigurationProxy::default()),
+            genesis: Some(genesis::ConfigurationProxy::default()),
+            wsv: Some(wsv::ConfigurationProxy::default()),
+            network: Some(network::ConfigurationProxy::default()),
+            telemetry: Some(telemetry::ConfigurationProxy::default()),
+        }
+    }
+}
+
+impl ConfigurationProxy {
+    /// Finalisation of Iroha config, i.e. in case where only some of mutually equivalent fields
+    /// were provided, instantiation of all the missing fields happens.
+    ///
+    /// # Errors
+    /// Fails early if none of the fields from the corresponding mutually equivalent set were
+    /// provided.
+    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    pub fn finalize(&mut self) -> Result<()> {
+        if let Some(sumeragi_proxy) = &mut self.sumeragi {
+            // First, iroha public/private key and sumeragi keypair are interchangeable, but
+            // the user is allowed to provide only the former, and keypair is generated automatically
+            if let (Some(public_key), Some(private_key)) = (&self.public_key, &self.private_key) {
+                sumeragi_proxy.key_pair =
+                    Some(KeyPair::new(public_key.clone(), private_key.clone())?);
+            } else {
+                eyre::bail!(ConfigError::ProxyBuildError(
+                    "Iroha public and private key not supplied, instantiating sumeragi keypair is impossible"
+                        .to_owned()
+                ))
+            }
+            // Second, torii gateway and sumeragi peer id are interchangeable too
+            if let Some(torii_proxy) = &mut self.torii {
```

Comment:
I want emphasize on this idea one more time. 

If we _can_ construct `sumeragi.peer_id` from `public_key` and `torii.p2p_addr` then we probably _should_ construct it from this values. It's invalid state to have `peer_id.addr != p2p_addr` or `peer.public_key != public_key`.

I believe that we avoid a lot pain in the ass if we would restrict configuration to eliminate possible divergence of configuration parameters.